File: AlgorithmJavaVersion/src/DataStructure/sort/innerSort/innerSortImpl/HeapSort2.java
package DataStructure.sort.innerSort.innerSortImpl;

import Common.Utils.UTFactory;
import org.junit.Test;
import DataStructure.sort.innerSort.ArraySortINTF;

/**
 * @author liujun
 * @version 1.0
 * @date 2020/6/21
 * @author—Email ljfirst@mail.ustc.edu.cn
 * @blogURL https://blog.csdn.net/ljfirst
 * @description 堆排序
 */
public class HeapSort2 implements ArraySortINTF {

    @Test // 验证功能：用于全量测试
    public void TestFunc() throws Exception {
        UTFactory.FullTest(this.getClass());
    }

    @Test // 调试功能 : 用于复现错误测试案例
    public void TestDoubleTrack() throws Exception {
        String input = "";
        String output = "";
        UTFactory.DebugTest(this.getClass(), input, output);
    }

    
    public void sortMethod(int[] array) {
        //输入检查
        if (!check(array)) {
            return;
        }
        int length = array.length - 1;
        for (int i = 0; i < length; i++) {
            heapify_big(array, length - i);
            swap(array, 0, length - i);
        }
    }

    //小顶堆
    public void heapify_big(int[] array, int border) {

        /**
         * 最后一个非叶子结点的index见 {@link HeapSort#sortMethod},java grammar: @link的用法：类名#方法或者属性
         * 参考blog：https://blog.csdn.net/qq_27093465/article/details/59121608
         */
        for (int root = (border - 1) / 2; root >= 0; root--) {
            int child = 2 * root + 1;
            // attention:边界值的 等于 需要特别注意
            if (child + 1 <= border) {
                child = array[child] > array[child + 1] ? child : child + 1;
            }
            if (child <= border && array[child] > array[root]) {
                swap(array, child, root);
            }
        }
    }
}

File: AlgorithmGoVersion/src/DataStructure/sort/innerSort/innerSortImpl/HeapSort2.go
package innerSortImpl

import (
	"AlgorithmPractice/src/Common/Utils"
)

/**
 * @author 蔚蔚樱
 * @version 1.0
 * @date 2021-06-22 00:09
 * @author—Email 蔚蔚樱first@outlook.com
 * @blogURL https://blog.csdn.net/ljfirst
 * @description 堆排序 2
 */

type HeapSort2 struct {
}

func (s *HeapSort2) SortMethod(array []int) []int {
	if array == nil || len(array) <= 1 {
		return array
	}

	for i := 0; i < len(array)-1; i++ {
		//整堆
		s.heapify2_big(array, len(array)-1-i)
		//交换
		array[0], array[len(array)-1-i] = array[len(array)-1-i], array[0]
	}
	return array
}

func (s *HeapSort2) heapify2_big(array []int, border int) {

	for root := (border - 1) / 2; root >= 0; root-- {
		child := root*2 + 1
		if child > border {
			break
		}
		// tips002:边界值的等于需要特别注意
		if child+1 <= border {
			child = Utils.GetArrayMaxIndex(array, child, child+1)
		}
		if array[child] > array[root] {
			array[child], array[root] = array[root], array[child]
		}
	}
}

func (s *HeapSort2) InterfaceName() string {
	return "ArraySortINTF"
}